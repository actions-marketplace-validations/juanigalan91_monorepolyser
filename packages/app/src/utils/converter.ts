import { NODE_TYPES, Node, Graph } from '../types';

const graph: Graph = {
    nodes: [],
    edges: [],
    nodesIdsByLabel: {},
    totalNodes: 0,
}

const createNode = (id: number, label: string, type: NODE_TYPES): Node => {
    const node: Node = {
        id: id,
        label: label,
        shape: 'box',
    };

    switch(type) {
        case NODE_TYPES.TOP_DEPENDENCY:
            Object.assign(node, {
                color: {
                    background: '#28336d',
                },
                font: {
                    color: '#FFF',
                },
                scaling: {
                    label: {
                        enabled: true,
                    }
                },
            });
        break;
        case NODE_TYPES.DEV_DEPENDENCY:
            Object.assign(node, {
                color: '#5ac085',
            });
        break;
        case NODE_TYPES.DEPENDENCY:
            Object.assign(node, {
                color: '#eaf4ff',
            });
        break;
    }

    return node;
}

const addNode = (label: string, type: NODE_TYPES) => {
    let addedNodeId;

    if (!graph.nodesIdsByLabel[label]) {
        graph.totalNodes = graph.totalNodes + 1;
        graph.nodesIdsByLabel[label] = graph.totalNodes;

        graph.nodes.push(createNode(graph.totalNodes, label, type));

        addedNodeId = graph.totalNodes;
    } else {
        addedNodeId = graph.nodesIdsByLabel[label];
    }

    return addedNodeId;
}

const addNodes = (currentNodeId: number, rawNodes: Record<string, string>, type: NODE_TYPES) => {
    if (rawNodes) {
        const nodesKeys = Object.keys(rawNodes);

        nodesKeys.forEach((node: string) => {
            const depNodeId = addNode(node, type);

            graph.edges.push({
                from: currentNodeId,
                to: depNodeId,
            });
        });
    }
}

const convert = (rawData: any) => {
    const keys = Object.keys(rawData);
    let currentNodeId: number;

    keys.forEach((key) => {
        currentNodeId = addNode(key, NODE_TYPES.TOP_DEPENDENCY);

        const { dependencies, devDependencies } = rawData[key];

        addNodes(currentNodeId, dependencies, NODE_TYPES.DEPENDENCY);
        addNodes(currentNodeId, devDependencies, NODE_TYPES.DEV_DEPENDENCY)
    });

    return graph;
};

export { convert };
