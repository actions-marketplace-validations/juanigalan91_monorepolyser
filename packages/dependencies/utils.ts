import * as path from 'path';
import { Body } from 'package-json-types';

/**
 * Retrieves the workspaces property from a package json
 * @param packageJson - package to retrieve the workspaces from
 */
const getWorkspacesFromPackageJson = (packageJson: Body): Body['workspaces'] | null => {
  const { workspaces } = packageJson;

  if (Array.isArray(workspaces)) {
    return workspaces;
  }

  return null;
};

/** Retrieves the current working directory for the target repo */
const getWorkingDirectory = (): string => {
  return process.env.GITHUB_WORKSPACE ? process.env.GITHUB_WORKSPACE : process.cwd();
};

/** Returns the package json located on the root of the target repo */
const getRootPackageJson = (): Body => {
  const root = getWorkingDirectory();

  // eslint-disable-next-line import/no-dynamic-require, global-require
  return require(path.join(root, 'package.json'));
};

/** Returns the workspaces property located on the root of the target repo */
const getWorkspaces = (): Body['workspaces'] => {
  const workspaces = getWorkspacesFromPackageJson(getRootPackageJson());

  return workspaces;
};

export { getWorkspacesFromPackageJson, getRootPackageJson, getWorkingDirectory, getWorkspaces };
