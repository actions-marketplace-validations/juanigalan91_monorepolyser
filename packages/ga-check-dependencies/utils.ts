import { getWorkspaces, getPackagesFromWorkspaces } from '@monorepolyser/dependencies';

import { RepeatedDependency } from './types';

export interface GetIncoherentDependencies {
  incoherentDependencies: Record<string, RepeatedDependency[]>;
  deps: Record<string, string>;
}

const getIncoherentDependencies = (): GetIncoherentDependencies => {
  const incoherentDependencies: Record<string, RepeatedDependency[]> = {};
  const deps: Record<string, string> = {};
  const workspaces = getWorkspaces();
  const { packages } = getPackagesFromWorkspaces(workspaces);

  Object.keys(packages).forEach((pkgName) => {
    const pkg = packages[pkgName];
    const { dependencies } = pkg;

    if (dependencies) {
      Object.keys(dependencies).forEach((dep) => {
        const version = dependencies[dep];

        if (deps[dep]) {
          const detectedVersion = deps[dep];

          if (version !== detectedVersion) {
            if (!incoherentDependencies[dep]) {
              incoherentDependencies[dep] = [];
            }

            incoherentDependencies[dep].push({
              addedBy: pkgName,
              version,
            });
          }
        } else {
          deps[dep] = version;
        }
      });
    }
  });

  return {
    incoherentDependencies,
    deps,
  };
};

export { getIncoherentDependencies };
