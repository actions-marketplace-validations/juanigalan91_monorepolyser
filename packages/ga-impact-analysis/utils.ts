import { ProjectMetadata } from '@monorepolyser/dependencies/types';

const calculatePackagesDependencies = (project: ProjectMetadata) => {
  const dependedOnPackages: Record<string, number> = {};
  const { packages } = project;

  Object.keys(packages).forEach((pkgName) => {
    const pkg = packages[pkgName];
    const { dependencies } = pkg;

    if (dependencies) {
      Object.keys(dependencies).forEach((dep) => {
        if (dependedOnPackages[dep]) {
          dependedOnPackages[dep] += 1;
        } else {
          dependedOnPackages[dep] = 1;
        }
      });
    }
  });

  return { dependedOnPackages };
};

export { calculatePackagesDependencies };
