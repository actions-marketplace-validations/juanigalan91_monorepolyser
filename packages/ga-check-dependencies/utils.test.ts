import * as utils from './utils';

describe('Check Dependencies - Utils', () => {
  it('returns no incoherent dependencies when all packages use the same versions', () => {
    const project = {
      packages: {
        '@module/admin': {
          dependencies: {
            react: '16.8.3',
          },
          name: '@module/admin',
          version: '1.0.0',
        },
        '@module/home': {
          dependencies: {
            lodash: '0.19.0',
            react: '16.8.3',
          },
          name: '@module/home',
          version: '1.0.0',
        },
        '@module/root': {
          dependencies: {
            jest: '26.8.3',
          },
          name: '@module/root',
          version: '1.0.0',
        },
        '@module/search': {
          name: '@module/search',
          version: '1.0.0',
        },
      },
      packagesByPath: {},
      workspaces: ['packages/**'],
    };

    const deps = utils.getIncoherentDependencies(project);

    expect(deps).toMatchSnapshot();
    expect(Object.keys(deps.incoherentDependencies)).toHaveLength(0);
    expect(Object.keys(deps.deps)).toHaveLength(3);
  });

  it('returns incoherent dependencies when a package is using a different version', () => {
    const project = {
      packages: {
        '@module/admin': {
          dependencies: {
            react: '16.8.3',
          },
          name: '@module/admin',
          version: '1.0.0',
        },
        '@module/home': {
          dependencies: {
            lodash: '0.19.0',
            react: '16.8.4',
          },
          name: '@module/home',
          version: '1.0.0',
        },
        '@module/root': {
          dependencies: {
            jest: '26.8.3',
          },
          name: '@module/root',
          version: '1.0.0',
        },
        '@module/search': {
          name: '@module/search',
          version: '1.0.0',
        },
      },
      packagesByPath: {},
      workspaces: ['packages/**'],
    };

    const deps = utils.getIncoherentDependencies(project);
    const incohorentDeps = Object.keys(deps.incoherentDependencies);
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    const reactDep = deps.incoherentDependencies.react;

    expect(deps).toMatchSnapshot();
    expect(incohorentDeps).toHaveLength(1);
    expect(reactDep).toHaveLength(1);
    expect(reactDep[0].addedBy).toEqual('@module/home');
    expect(Object.keys(deps.deps)).toHaveLength(3);
  });

  it('returns incoherent dependencies when a package is using several different versions', () => {
    const project = {
      packages: {
        '@module/admin': {
          dependencies: {
            react: '16.8.3',
          },
          name: '@module/admin',
          version: '1.0.0',
        },
        '@module/home': {
          dependencies: {
            lodash: '0.19.0',
            react: '16.8.4',
          },
          name: '@module/home',
          version: '1.0.0',
        },
        '@module/root': {
          dependencies: {
            jest: '26.8.3',
          },
          name: '@module/root',
          version: '1.0.0',
        },
        '@module/search': {
          name: '@module/search',
          version: '1.0.0',
          dependencies: {
            react: '16.8.5',
          },
        },
      },
      packagesByPath: {},
      workspaces: ['packages/**'],
    };

    const deps = utils.getIncoherentDependencies(project);
    const incohorentDeps = Object.keys(deps.incoherentDependencies);
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    const reactDep = deps.incoherentDependencies.react;

    expect(deps).toMatchSnapshot();
    expect(incohorentDeps).toHaveLength(1);
    expect(reactDep).toHaveLength(2);
    expect(reactDep[0].addedBy).toEqual('@module/home');
    expect(Object.keys(deps.deps)).toHaveLength(3);
  });
});
