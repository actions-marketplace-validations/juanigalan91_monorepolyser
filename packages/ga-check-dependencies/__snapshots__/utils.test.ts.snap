// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Dependencies - Utils returns incoherent dependencies when a package is using a different version 1`] = `
Object {
  "deps": Object {
    "jest": "26.8.3",
    "lodash": "0.19.0",
    "react": "16.8.3",
  },
  "incoherentDependencies": Object {
    "react": Array [
      Object {
        "addedBy": "@module/home",
        "name": "react",
        "version": "16.8.4",
      },
    ],
  },
}
`;

exports[`Check Dependencies - Utils returns incoherent dependencies when a package is using several different versions 1`] = `
Object {
  "deps": Object {
    "jest": "26.8.3",
    "lodash": "0.19.0",
    "react": "16.8.3",
  },
  "incoherentDependencies": Object {
    "react": Array [
      Object {
        "addedBy": "@module/home",
        "name": "react",
        "version": "16.8.4",
      },
      Object {
        "addedBy": "@module/search",
        "name": "react",
        "version": "16.8.5",
      },
    ],
  },
}
`;

exports[`Check Dependencies - Utils returns no incoherent dependencies when all packages use the same versions 1`] = `
Object {
  "deps": Object {
    "jest": "26.8.3",
    "lodash": "0.19.0",
    "react": "16.8.3",
  },
  "incoherentDependencies": Object {},
}
`;
